pydantic/_internal/_typing_extra.py

```diff
@@ -245,50 +282,66 @@ def get_cls_type_hints(
         with ns_resolver.push(base):
             globalns, localns = ns_resolver.types_namespace
             for name, value in ann.items():
-                hints[name] = eval_type(value, globalns, localns, lenient=lenient)
+                if lenient:
+                    hints[name] = try_eval_type(value, globalns, localns)
+                else:
+                    hints[name] = eval_type(value, globalns, localns)
     return hints
 
 
-def eval_type(
+def try_eval_type(
     value: Any,
     globalns: GlobalsNamespace | None = None,
     localns: MappingNamespace | None = None,
-    *,
-    lenient: bool = False,
-) -> Any:
-    """Evaluate the annotation using the provided namespaces.
+) -> tuple[Any, bool]:
+    """Try evaluating the annotation using the provided namespaces.
 
     Args:
         value: The value to evaluate. If `None`, it will be replaced by `type[None]`. If an instance
             of `str`, it will be converted to a `ForwardRef`.
         localns: The global namespace to use during annotation evaluation.
         globalns: The local namespace to use during annotation evaluation.
-        lenient: Whether to keep unresolvable annotations as is or re-raise the `NameError` exception. Default: re-raise.
+
+    Returns:
+        A two-tuple containing the possibly evaluated type and a boolean indicating
+            whether the evaluation succeeded or not.
     """
-    if value is None:
-        value = NoneType
-    elif isinstance(value, str):
-        value = _make_forward_ref(value, is_argument=False, is_class=True)
+    value = _type_convert(value)
 
     try:
-        return eval_type_backport(value, globalns, localns)
+        return eval_type_backport(value, globalns, localns), True
     except NameError:
-        if not lenient:
-            raise
-        # the point of this function is to be tolerant to this case
-        return value
+        return value, False
+
+
+def eval_type(
+    value: Any,
+    globalns: GlobalsNamespace | None = None,
+    localns: MappingNamespace | None = None,
+) -> Any:
+    """Evaluate the annotation using the provided namespaces.
+
+    Args:
+        value: The value to evaluate. If `None`, it will be replaced by `type[None]`. If an instance
+            of `str`, it will be converted to a `ForwardRef`.
+        localns: The global namespace to use during annotation evaluation.
+        globalns: The local namespace to use during annotation evaluation.
+    """
+    value = _type_convert(value)
+    return eval_type_backport(value, globalns, localns)
 
 
 @deprecated(
-    '`eval_type_lenient` is deprecated, use `eval_type` with `lenient=True` instead.',
+    '`eval_type_lenient` is deprecated, use `try_eval_type` instead.',
```

Comment:
Hmm, this feels like an anti-pattern - if we already deprecated this, there's probably a reason why we didn't want to have two functions. Even if we disagree with this deprecation now, seems like a lot of churn to basically reintroduce the function and redirect for the deprecated case...

I believe my suggestion above re always returning a second value, `evaluated`, even in the case of `eval_type` with `lenient=False` might help, and we can leave this be?

Reply:
A few more follow ups:
1. This is a very internal function - can we just deprecate this?
2. If we were to bring back a second function, to reduce churn, can we call it `eval_type_lenient`?

Reply:
I had to keep this function during the types namespaces refactor because fastapi makes use of it :face_with_diagonal_mouth:. I added the `deprecated` decorator so that type checking fails in the fastapi CI and they can use the correct function (or best, do not rely on our internals because I don't think the `_typing_extra` module is stable enough, we might apply other changes here in the future).

> I believe my suggestion above re always returning a second value, `evaluated`, even in the case of `eval_type` with `lenient=False` might help, and we can leave this be?

Same issue as https://github.com/pydantic/pydantic/pull/10769/files#r1832947088, we can't have a single function because we want to catch the `NameError` to reraise `PydanticUndefinedAnnotation`